homeassistant/components/camera/push.py

```diff
@@ -0,0 +1,162 @@
+        cv.time_period, cv.positive_timedelta),
+    vol.Optional(CONF_IMAGE_FIELD, default='image'): cv.string,
+})
+
+
+async def async_setup_platform(hass, config, async_add_devices,
+                               discovery_info=None):
+    """Set up the Push Camera platform."""
+    cameras = [PushCamera(config[CONF_NAME],
+                          config[CONF_BUFFER_SIZE],
+                          config[CONF_TIMEOUT])]
+
+    hass.http.register_view(CameraPushReceiver(cameras,
+                                               config[CONF_IMAGE_FIELD]))
+
+    async_add_devices(cameras)
+
+
+class CameraPushReceiver(HomeAssistantView):
+    """Handle pushes from remote camera."""
+
+    url = "/api/camera_push/{entity_id}"
+    name = 'api:camera_push:camera_entity'
+
+    def __init__(self, cameras, image_field):
+        """Initialize CameraPushReceiver with camera entity."""
+        self._cameras = cameras
+        self._image = image_field
+
+    async def post(self, request, entity_id):
+        """Accept the POST from Camera."""
+        try:
+            (_camera,) = [camera for camera in self._cameras
+                          if camera.entity_id == entity_id]
+        except ValueError:
+            _LOGGER.error("Unknown push camera %s", entity_id)
+            return self.json_message('Unknown Push Camera',
+                                     HTTP_BAD_REQUEST)
+
+        try:
+            data = await request.post()
+            _LOGGER.debug("Received Camera push: %s", data[self._image])
+            await _camera.update_image(data[self._image].file.read(),
+                                       data[self._image].filename)
+        except ValueError as value_error:
+            _LOGGER.error("Unknown value %s", value_error)
+            return self.json_message('Invalid POST', HTTP_BAD_REQUEST)
+        except KeyError as key_error:
+            _LOGGER.error('In your POST message %s', key_error)
+            return self.json_message('{} missing'.format(self._image),
+                                     HTTP_BAD_REQUEST)
+
+
+class PushCamera(Camera):
+    """The representation of a Push camera."""
+
+    def __init__(self, name, buffer_size, timeout):
+        """Initialize push camera component."""
+        super().__init__()
+        self._name = name
+        self._last_trip = None
+        self._filename = None
+        self._expired = None
+        self._state = STATE_IDLE
+        self._timeout = timeout
+        self.queue = deque([], buffer_size)
+        self._current_image = None
+
+    @property
+    def state(self):
+        """Current state of the camera."""
+        return self._state
+
+    async def update_image(self, image, filename):
+        """Update the camera image."""
+        if self._state == STATE_IDLE:
+            self._state = STATE_RECORDING
+            self._last_trip = dt_util.utcnow()
+            self.queue.clear()
+
+        self._filename = filename
```

Comment:
do we really need this? That make a state change any new picture is pushed

Reply:
That is actually a feature :)

That state change will help automations that require a trigger (e.g. do some image_processing, send a notification).

Reply:
Yeah, that will be trigger if the state of camera going to recording, why you need after that a state change?

Reply:
In the use case of motioneye, an event will consist of 3 images (default)

I expect to be notified of the 3 for the aforementioned automations.

Reply:
I have a idea. Add the options `force_update` and use this on entity as attribute `force_update`.
Other platform make the same. That allow user to force state change on if a update is exists. I think it is a bit wired to store a filename into attribute to have the same effect. What do you think?

Reply:
For what? Can you give a example why to you need this attribute?

Reply:
Filename is a unique key, and can be used to trace events.

Personal example:
MotionEye sends HA an image, HA sends a notification with the image. If I want to go back to motionEye image storage, I need a key to look for. Sending the filename with the notification lets me go back to motionEye storage and look for the image and review the security event (way past the time HA has cleared the image from the queue)

Reply:
Ok. Remove the force_update if you want staying with that attribute and we can merge it.

Reply:
Done